docs/runtime-code-basics.md

```diff
@@ -0,0 +1,229 @@
+  nk.create_leaderboard(id, "desc", "0 0 * * 1", json, false)
+
+  return nk.json_encode({["id"] = id})
+  -- will return "{'id': 'some UUID'}" (JSON) as bytes
+end
+
+nk.register_rpc(some_example, "my_unique_id")
+```
+
+We import the `"nakama"` module which is embedded within the server and contains lots of server-side functions which are helpful as you build your code. You can see all available functions in the [module reference](runtime-code-function-reference.md).
+
+## Register hooks
+
+The code in a module will be evaluated immediately and can be used to register functions which can operate on messages from clients as well as execute logic on demand.
+
+All registered functions receive a "context" table as the first argument and "payload" as the second. The "context" contains fields which depend on when the code is executed.
+
+| Field | Purpose |
+| ----- | ------- |
+| `context.Env` | A table of key/value pairs which are defined in the YAML [configuration](install-configuration.md) of the server. This is useful to store API keys and other secrets which may be different between servers run in production and in development. |
+| `context.ExecutionMode` | The mode associated with the execution context. It's one of these values: "after", "before", "http", or "rpc". |
+| `context.UserHandle` | The user handle associated with the execution context. It will always be `nil` on `register_http`. |
+| `context.UserId` | The user ID associated with the execution context. It will always be `nil` on `register_http`. |
+| `context.UserSessionExp` | The user session expiry in milliseconds associated with the execution context. It will always be `nil` on `register_http`. |
+
+There are four ways to register a function within the runtime each of which is used to handle specific behavior between client and server.
+
+### register_before
+
+Any function may be registered to intercept a message received from a client and operate on it (or reject it) based on custom logic. This is useful to enforce specific rules on top of the standard features in the server.
+
+```lua hl_lines="10"
+local nk = require("nakama")
+
+local function limit_friends(context, payload)
+  local user = nk.user_fetch_id({context.UserId})[1]
+  -- lets assume we've stored a user's level in their metadata.
+  local metadata = nk.json_decode(user.Metadata)
+  if metadata.level <= 10 then
+    error("Must reach level 10 before you can add friends.")
+  end
+  return payload -- important!
+end
+nk.register_before(limit_friends, "TFriendAdd")
+```
+
+The code above fetches the current user's profile and checks the metadata which is assumed to be JSON encoded with `"{level: 12}"` in it. If a user's level is too low an error is thrown to prevent the Friend Add message from being passed onwards in the server pipeline.
+
+!!! Note
+    You must remember to pass onwards the payload to the rest of the message pipeline. See `"return payload"` highlighted in the code above.
```

Comment:
You must remember to return the payload at the end of your function in the same structure as you recieved it